tensorflow/core/lib/strings/str_util.cc

```diff
@@ -416,6 +417,103 @@ bool ConsumeNonWhitespace(StringPiece* s, StringPiece* val) {
   }
 }
 
+Status SplitUTF8(StringPiece text, const string& delim, const bool skip_empty,
+                 std::vector<string>* result) {
+  // Bytes    Byte 1    Byte 2    Byte 3    Byte 4
+  //   1     0xxxxxxx
+  //   2     110xxxxx  10xxxxxx
+  //   3     1110xxxx  10xxxxxx  10xxxxxx
+  //   4     11110xxx  10xxxxxx  10xxxxxx  10xxxxxx
+  result->clear();
+  size_t char_start = 0, char_len = 0;
+  bool advance = true;
+  for (size_t i = 0; i < text.size(); ++i) {
+    if (char_start == i) {
+      if (!UTF8CharNumBytes(text, &char_len)) {
+        result->clear();
+        return errors::InvalidArgument("Invalid UTF8 encoding at byte ", i);
+      }
+    }
+    // Validate all intermediate bytes are valid UTF-8
+    if (TF_PREDICT_FALSE((char_len > 1) && (char_start != i) &&
+                         (text[i] & 0xC0) != 0x80)) {
+      return errors::InvalidArgument("Invalid UTF8 encoding at byte ", i);
+    }
+    // Reached the end of a character
+    if (char_start + char_len == i + 1) {
+      const StringPiece entry = text.substr(char_start, char_len).ToString();
+      if (delim == "") {
+        result->emplace_back(entry.ToString());
+      } else if (delim == entry) {
+        advance = true;
+        if (!skip_empty) {
+          // Follow python style, if it is beginning or the end, always add ''
+          // >>> "##a##b##c##".split("#")
+          // ['', '', 'a', '', 'b', '', 'c', '', '']
+          if (result->size() == 0 || result->back() != "") {
+            result->emplace_back("");
+          } else if (result->size() == 1) {
```

Comment:
I didn't quite follow the logic here (in particulary while `result->size() == 1` is any more special that `result->size() == 2` etc.). For example, let's say the the call is:

```
SplitUTF8("###a#", '#')
```

then I think this will end up with: `{ "", "", "a", ""}` instead of three empty strings before "a".
If that is correct, probably worth adding the unittest that would have caught the bug here.

I think this can be simplified a bit by keeping a boolean indicating whether the last character was a delimiter and using that instead. Something like:

```c++
} else if (delim == entry) {
  if (last_char_was_a_delimiter && !skip_empty) {
    result->emplace_back("");
  }
  last_char_was_a_delimiter = true;
} else if (last_char_was_a_delimiter) {
  last_char_was_a_delimiter = false;
  result->emplace_back(entry.ToString());
} else {
  result->back() += enty.ToString();
}
```

Reply:
@asimshankar I think some special handling of begin and end may still be needed. For example, in python:
```python
Python 2.7.10 (default, Feb  7 2017, 00:08:15) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.34)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> "###a#".split("#")
['', '', '', 'a', '']
>>> ",".split(",")
['', '']
>>> 
```

The first case ` "###a#".split("#")` returns 3 empty chars before `a` and the second case will return 2 chars.

I updated the PR to try to match the above situation. Please take a look and see if it fit.